cpp/src/io/utilities/datasource.cpp

```diff
@@ -135,6 +137,65 @@ class memory_mapped_source : public datasource {
   size_t map_offset_ = 0;
 };
 
+/**
+ * @brief Implementation class for reading from a file optimized for reading
+ * to device memory. Attempts to hide HtoD transfer latency with disk io by using
+ * a double-buffer.
+ *
+ **/
+class gpu_async_filereader : public datasource {
+  static constexpr uint32_t default_buffer_size = 16 << 20; // 16MB
+  struct pinned_filebuf {
+    uint8_t *host_bfr = nullptr;
+    cudaStream_t stream = 0;
+    cudaEvent_t completion_event = 0;
+    uint32_t position = 0;
+    bool event_pending = false;
+    ~pinned_filebuf() {
+      if (completion_event) {
+        cudaEventDestroy(completion_event);
+      }
+      if (host_bfr) {
+        cudaFreeHost(host_bfr);
+      }
+    }
+  };
+ public:
+  explicit gpu_async_filereader(const std::string& filepath, uint32_t bfrsize = default_buffer_size):
+    fd_(open(filepath.c_str(), O_RDONLY)), bufid_(0), buffer_size_(bfrsize) {
+    CUDF_EXPECTS(fd_ != -1, "Cannot open file");
+  }
+  ~gpu_async_filereader() {
+    if (fd_ != -1) {
+      close(fd_);
+    }
+  }
+  size_t size() const override {
+    struct stat st {};
```

Comment:
This looks like C syntax.

```suggestion
    stat st {};
```

Reply:
Will keep that in mind ~ still very early WIP FWIW, likely to switch to arrow::RandomAccessFile in final implementation.

Reply:
It is in fact a C struct, and struct appears to be necessary.